lib/src/revset.rs

```diff
@@ -1785,7 +1950,8 @@ pub fn optimize<St: ExpressionState>(
     let expression = fold_generation(&expression).unwrap_or(expression);
     let expression = internalize_filter(&expression).unwrap_or(expression);
     let expression = fold_difference(&expression).unwrap_or(expression);
```

Comment:
it's probably better to insert `fold_heads_range()` between `internalize_filter()` and `fold_difference()`. Then, we wouldn't have to deal with compound expression nodes.